cpp/src/allgather/allgather.cpp

```diff
@@ -0,0 +1,567 @@
+            return std::move(chunk);
+        }
+    );
+    auto cit = chunks.begin() + static_cast<std::int64_t>(complete.size());
+    auto fit = futures.begin();
+    for (; cit != chunks.end() && fit != futures.end(); cit++, fit++) {
+        RAPIDSMPF_EXPECTS(*fit, "Not expecting nullptr here");
+        RAPIDSMPF_EXPECTS(!(*cit).second, "Expecting no future here");
+        std::swap((*cit).second, *fit);
+    }
+    auto osize = chunks.size();
+    std::erase(
+        chunks,
+        std::pair<std::unique_ptr<Chunk>, std::unique_ptr<Communicator::Future>>{
+            nullptr, nullptr
+        }
+    );
+    RAPIDSMPF_EXPECTS(chunks.size() == osize - result.size(), "Something went wrong");
+    return result;
+}
+}  // namespace detail
+
+class AllGather::Progress {
+  public:
+    Progress(AllGather& gather) : gather_{gather} {}
+
+    ProgressThread::ProgressState operator()() {
+        return gather_.event_loop();
+    }
+
+  private:
+    AllGather& gather_;
+};
+
+void AllGather::insert(PackedData&& packed_data) {
+    if (packed_data.data->size == 0) {
+        // No point communicating zero-sized insertions.
+        // Note: this means the caller must handle the metadata
+        // aspects of the case where the extraction from an allgather
+        // is empty.
+        return;
+    }
+    nlocal_insertions_.fetch_add(1, std::memory_order_relaxed);
+    return insert(
+        detail::Chunk::from_packed_data(
+            sequence_number_.fetch_add(1, std::memory_order_relaxed),
+            comm_->rank(),
+            std::move(packed_data)
+        )
+
+    );
+}
+
+void AllGather::insert(std::unique_ptr<detail::Chunk> chunk) {
+    RAPIDSMPF_EXPECTS(
+        !locally_finished_.load(std::memory_order_acquire),
+        "Can't insert after locally indicating finished"
+    );
+    inserted_.insert(std::move(chunk));
+}
+
+void AllGather::insert_finished() {
+    locally_finished_.store(true, std::memory_order_release);
+    inserted_.insert(
+        detail::Chunk::from_empty(
+            static_cast<std::uint32_t>(
+                nlocal_insertions_.load(std::memory_order_acquire)
+            ),
+            comm_->rank()
+        )
+    );
+}
+
+bool AllGather::finished() const noexcept {
+    return finish_counter_.load(std::memory_order_acquire) == 0
+           && for_extraction_.ready();
+}
+
+std::vector<PackedData> AllGather::wait_and_extract(AllGather::Ordered ordered) {
+    wait();
+    auto chunks = for_extraction_.extract();
+    std::vector<PackedData> result;
+    result.reserve(chunks.size());
+    if (ordered == AllGather::Ordered::YES) {
+        std::ranges::sort(chunks, std::less{}, [](auto&& chunk) { return chunk->id(); });
+    }
+    std::ranges::transform(chunks, std::back_inserter(result), [](auto&& chunk) {
+        return chunk->release();
+    });
+    return result;
+}
+
+std::vector<PackedData> AllGather::extract_ready() {
+    auto chunks = for_extraction_.extract();
```

Comment:
Is this correct? `extract()` says "Extract all chunks from the postbox." and not just ready chunks, do you need `extract_ready()` instead?

Reply:
Ah, so this is subtle and I should update the docs. Insertion to and extraction from an `AllGather` is promised to be stream-ordered wrt the stream the user provided to the constructor.

Buffers are moved into the `for_extraction_` postbox when their future completes (and so there are no pending communications on them). There _may_ (there actually aren't) be pending stream-ordered operations on the buffers, but that is OK, because we have promised to be stream ordered I think.

I suppose this might fall over for the case where a buffer has been asynchronously spilled from device to host though...